[DI] Overriding services autowired by name under _defaults bind not working

przemyslaw-bogusz opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #28326 
| License       | MIT

This is an implementation of ideas and suggestions of @nicolas-grekas and @GuilhemN.

przemyslaw-bogusz committed:
[DI] Fixes: #28326 - Overriding services autowired by name under _defaults bind not working

OskarStark commented:
Typo in PR headline:
```diff
- [DI] Fixes: #28326 - Overriding services autowired by name under _deaults bind not working
+ [DI] Fixes: #28326 - Overriding services autowired by name under _defaults bind not working
```

GuilhemN reviewed (commented):
This is great :+1: I'm only concerned about the `PhpFileLoader` which doesn't call `FileDefinition::setDefinition()` (https://github.com/przemyslaw-bogusz/symfony/blob/85a28f1e4e5a6197bbc4ed0550b9ce68936c83bf/src/Symfony/Component/DependencyInjection/Loader/Configurator/ServiceConfigurator.php#L65).
Shouldn't the logic be added there too?

nicolas-grekas commented:
See comment in #29949: there may be an issue that makes the container load all classes at compile time. We should avoid it if confirmed.

przemyslaw-bogusz commented:
@GuilhemN Thanks for the suggestion. I added the funcionality for `PhpFileLoader`, and as a consequence I moved code from `FileLoader` into `ContainerBuilder` to avoid duplication. 

@nicolas-grekas This PR is rather simple so I don't think I've made a change that would cause 

> the container load all classes at compile time

At least, I'm not aware of that.

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/ContainerBuilder.php:
```diff
@@ -1504,6 +1506,35 @@ public function normalizeId($id)
+    /**
+     * Gets removed binding ids.
+     *
+     * @return array
```
please keep only `@internal` in the docblock (no title) and use a real return type

OskarStark replied:
Are you sure with the return type, it’s against 3.4 branch, or am I mixing things up?

nicolas-grekas replied:
oh, 3.4, my bad!

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/ContainerBuilder.php:
```diff
@@ -1504,6 +1506,35 @@ public function normalizeId($id)
+     *
+     * @internal
+     */
+    public function addRemovedBindingIds($id)
```
`int` instead of annotation+docblock title

teohhanhui replied:
This should actually be `string`.

nicolas-grekas replied:
good catch, could you please send a PR to fix it?

teohhanhui replied:
https://github.com/symfony/symfony/pull/31173

nicolas-grekas reviewed (commented):
cannot test right now,here is a very light review

Commit:
more tests

nicolas-grekas commented:
Thank you @przemyslaw-bogusz.
